Decorating QuarkusApplication won't build

## Describe the bug
Creating a decorator for the `QuarkusApplication` interface gives a build error about only decorators can declare a delegate (even if the decorator is properly marked as `@Decorator`)

### Expected behavior
According to conversation on Zulip, this should work, so the app should build and the whole Quarkus execution decorated

### Actual behavior
Build error:
```java
[ERROR] Failed to execute goal io.quarkus:quarkus-maven-plugin:2.0.0.CR3:build (default) on project app: Failed to build quarkus application: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
[ERROR] 	[error]: Build step io.quarkus.arc.deployment.ArcProcessor#validate threw an exception: javax.enterprise.inject.spi.DeploymentException: Only decorators can declare a delegate injection point: CLASS bean [types=[io.quarkus.runtime.QuarkusApplication, java.lang.Object, AppDecorator], qualifiers=[@Default, @Any], target=AppDecorator]
[ERROR] 	at io.quarkus.arc.processor.BeanInfo.init(BeanInfo.java:457)
```

## To Reproduce

Reproducer: [main-decorator.zip](https://github.com/quarkusio/quarkus/files/6726011/main-decorator.zip)

Steps to reproduce the behavior:
1. Just build with Maven

## Environment (please complete the following information):

### Output of `uname -a` or `ver`
`Darwin xtaixe 20.5.0 Darwin Kernel Version 20.5.0: Sat May  8 05:10:33 PDT 2021; root:xnu-7195.121.3~9/RELEASE_X86_64 x86_64`

### Output of `java -version`
```
openjdk version "11.0.11" 2021-04-20
OpenJDK Runtime Environment GraalVM CE 21.1.0 (build 11.0.11+8-jvmci-21.1-b05)
OpenJDK 64-Bit Server VM GraalVM CE 21.1.0 (build 11.0.11+8-jvmci-21.1-b05, mixed mode, sharing)
```

### Quarkus version or git rev
`2.0.0.CR3`

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)
```
Apache Maven 3.8.1 (05c21c65bdfed0f71a2f2ada8b84da59348c4c5d)
Maven home: /Users/xtaixe/.mvnvm/apache-maven-3.8.1
Java version: 11.0.11, vendor: GraalVM Community, runtime: /Library/Java/JavaVirtualMachines/graalvm-ce-java11-21.1.0/Contents/Home
Default locale: en_US, platform encoding: UTF-8
OS name: "mac os x", version: "11.4", arch: "x86_64", family: "mac"
```

--- a/extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/ArcProcessor.java
+++ b/extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/ArcProcessor.java
@@ -116,12 +116,17 @@ public class ArcProcessor {
     }
 
     @BuildStep
-    AdditionalBeanBuildItem quarkusApplication(CombinedIndexBuildItem combinedIndexBuildItem) {
+    AdditionalBeanBuildItem quarkusApplication(CombinedIndexBuildItem combinedIndex) {
+        List<String> quarkusApplications = new ArrayList<>();
+        for (ClassInfo quarkusApplication : combinedIndex.getIndex()
+                .getAllKnownImplementors(DotName.createSimple(QuarkusApplication.class.getName()))) {
+            if (quarkusApplication.classAnnotation(DotNames.DECORATOR) == null) {
+                quarkusApplications.add(quarkusApplication.name().toString());
+            }
+        }
         return AdditionalBeanBuildItem.builder().setUnremovable()
                 .setDefaultScope(DotName.createSimple(ApplicationScoped.class.getName()))
-                .addBeanClasses(combinedIndexBuildItem.getIndex()
-                        .getAllKnownImplementors(DotName.createSimple(QuarkusApplication.class.getName())).stream()
-                        .map(s -> s.name().toString()).toArray(String[]::new))
+                .addBeanClasses(quarkusApplications)
                 .build();
     }
 

--- a/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanDeployment.java
+++ b/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanDeployment.java
@@ -773,8 +773,9 @@ public class BeanDeployment {
                 continue;
             }
 
-            if (annotationStore.hasAnnotation(beanClass, DotNames.INTERCEPTOR)) {
-                // Skip interceptors
+            if (annotationStore.hasAnnotation(beanClass, DotNames.INTERCEPTOR)
+                    || annotationStore.hasAnnotation(beanClass, DotNames.DECORATOR)) {
+                // Skip interceptors and decorators
                 continue;
             }
 

--- a/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/Decorators.java
+++ b/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/Decorators.java
@@ -49,6 +49,11 @@ final class Decorators {
                 priority = annotation.value().asInt();
                 priorityDeclared = true;
             }
+            ScopeInfo scopeAnnotation = beanDeployment.getScope(annotation.name());
+            if (scopeAnnotation != null && !BuiltinScope.DEPENDENT.is(scopeAnnotation)) {
+                throw new DefinitionException(
+                        "A decorator must be @Dependent but " + decoratorClass + " declares " + scopeAnnotation.getDotName());
+            }
         }
 
         //  The set includes all bean types which are Java interfaces, except for java.io.Serializable

--- a/independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/decorators/SimpleDecoratorTest.java
+++ b/independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/decorators/SimpleDecoratorTest.java
@@ -8,6 +8,7 @@ import javax.annotation.Priority;
 import javax.decorator.Decorator;
 import javax.decorator.Delegate;
 import javax.enterprise.context.ApplicationScoped;
+import javax.enterprise.context.Dependent;
 import javax.inject.Inject;
 import org.junit.jupiter.api.Test;
 import org.junit.jupiter.api.extension.RegisterExtension;
@@ -45,6 +46,7 @@ public class SimpleDecoratorTest {
 
     }
 
+    @Dependent
     @Priority(1)
     @Decorator
     static class TrimConverterDecorator implements Converter<String> {

--- a/independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/decorators/validation/NonDependentScopeTest.java
+++ b/independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/decorators/validation/NonDependentScopeTest.java
@@ -1,0 +1,53 @@
+package io.quarkus.arc.test.decorators.validation;
+
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+import io.quarkus.arc.test.ArcTestContainer;
+import javax.annotation.Priority;
+import javax.decorator.Decorator;
+import javax.decorator.Delegate;
+import javax.enterprise.context.ApplicationScoped;
+import javax.inject.Inject;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+public class NonDependentScopeTest {
+
+    @RegisterExtension
+    public ArcTestContainer container = ArcTestContainer.builder()
+            .beanClasses(Converter.class, DecoratorWithWrongScope.class).shouldFail().build();
+
+    @Test
+    public void testFailure() {
+        assertNotNull(container.getFailure());
+        assertTrue(
+                container.getFailure().getMessage()
+                        .contains(
+                                "A decorator must be @Dependent but io.quarkus.arc.test.decorators.validation.NonDependentScopeTest$DecoratorWithWrongScope declares javax.enterprise.context.ApplicationScoped"),
+                container.getFailure().getMessage());
+    }
+
+    interface Converter<T, U> {
+
+        T convert(T value);
+
+    }
+
+    @ApplicationScoped
+    @Priority(1)
+    @Decorator
+    static class DecoratorWithWrongScope implements Converter<String, String> {
+
+        @Inject
+        @Delegate
+        Converter<String, String> delegate;
+
+        @Override
+        public String convert(String value) {
+            return null;
+        }
+
+    }
+
+}

